storage/remote/queue_manager.go

```diff
@@ -266,6 +286,44 @@ func (t *QueueManager) Stop() {
 	level.Info(t.logger).Log("msg", "Remote storage stopped.")
 }
 
+func (t *QueueManager) Name() string {
+	return t.queueName
+}
+
+// Keep track of which series we know about for lookups when sending samples to remote.
+func (t *QueueManager) StoreSeries(series []tsdb.RefSeries) {
+	temp := make(map[uint64][]prompb.Label)
+	for i := 0; i < len(series); i++ {
+		var ls = make(model.LabelSet)
+		for _, label := range series[i].Labels {
+			ls[model.LabelName(label.Name)] = model.LabelValue(label.Value)
+		}
+		t.processExternalLabels(ls)
+		relabel.Process(ls, t.relabelConfigs...)
+		temp[series[i].Ref] = labelsToLabelsProto(labelsetToLabels(ls))
+	}
+	t.seriesMtx.Lock()
+	defer t.seriesMtx.Unlock()
+	for ref, labels := range temp {
+		t.series[ref] = labels
+	}
+}
+
+// Clear all series stored in series labels cache.
+func (t *QueueManager) ClearSeries() {
```

Comment:
I think this could race with already enqueued samples; before you clear the series, you need to ensure everything has been flushed.

Reply:
Or you could look up the labels for a series in `Append`, which is called on the same goroutine as this so can't race.

Reply:
samples are now enqueued with their labels